packages/idempotency/src/persistence/DynamoDbPersistenceLayer.ts

```diff
@@ -1,18 +1,73 @@
+  private _table: DynamoDBDocument | undefined;
+
+  public constructor(private tableName: string, private key_attr: string = 'id',
+    private status_attr: string = 'status', private expiry_attr: string = 'expiration',
+    private in_progress_expiry_attr: string = 'in_progress_expiry_attr',
+    private data_attr: string = 'data') {
     super();
   }
-  protected async _deleteRecord(): Promise<void> {}
-  protected async _getRecord(): Promise<IdempotencyRecord> {
-    return Promise.resolve({} as IdempotencyRecord);
+
+  protected async _deleteRecord(record: IdempotencyRecord): Promise<void> {
+    const table: DynamoDBDocument = this.getTable();
+    await table.delete({
+      TableName: this.tableName, Key: { [this.key_attr]: record.idempotencyKey }
+    });
+  }
+
+  protected async _getRecord(idempotencyKey: string): Promise<IdempotencyRecord> {
+    const table: DynamoDBDocument = this.getTable();
+    const output: GetCommandOutput = await table.get(
+      {
+        TableName: this.tableName, Key: { [this.key_attr]: idempotencyKey }
+      }
+    );
+
+    if (!output.Item) {
+      throw new IdempotencyItemNotFoundError();
+    }
+
+    return new IdempotencyRecord(output.Item[this.key_attr], output.Item[this.status_attr], output.Item[this.expiry_attr], output.Item[this.in_progress_expiry_attr], output.Item[this.data_attr], undefined);
+  }
+
+  protected async _putRecord(_record: IdempotencyRecord): Promise<void> {
+    const table: DynamoDBDocument = this.getTable();
+
+    const item = { [this.key_attr]: _record.idempotencyKey, [this.expiry_attr]: _record.expiryTimestamp, [this.status_attr]: _record.getStatus() };
+
+    const idempotencyKeyDoesNotExist = 'attribute_not_exists(#id)';
+    const idempotencyKeyExpired = '#expiry < :now';
+    const notInProgress = 'NOT #status = :inprogress';
+    const conditionalExpression = `${idempotencyKeyDoesNotExist} OR ${idempotencyKeyExpired} OR ${notInProgress}`;
+    await table.put({ TableName: this.tableName, Item: item, ExpressionAttributeNames: { '#id': this.key_attr, '#expiry': this.expiry_attr, '#status': this.status_attr }, ExpressionAttributeValues: { ':now': Date.now(), ':inprogress': IdempotencyRecordStatus.INPROGRESS }, ConditionExpression: conditionalExpression });
+  }
+
+  protected async _updateRecord(record: IdempotencyRecord): Promise<void> {
+    const table: DynamoDBDocument = this.getTable();
+    await table.update(
+      {
+        TableName: this.tableName, Key: { [this.key_attr]: record.idempotencyKey },
+        UpdateExpression: 'SET #status = :status, #expiry = :expiry', ExpressionAttributeNames: { '#status': this.status_attr, '#expiry': this.expiry_attr }, ExpressionAttributeValues: { ':status': record.getStatus(), ':expiry': record.expiryTimestamp }
+      }
+    );
+  }
+
+  private getTable(): DynamoDBDocument {
+    if (!this._table)
+      this._table = DynamoDBDocument.from(new DynamoDB({}));
```

Comment:
I just want to make sure that if a record field like "expiryTimestamp" is undefined, that it won't blow up during the implicit marshalling that lib-dynamodb does

https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/modules/_aws_sdk_lib_dynamodb.html#configuration

Reply:
Are you suggesting then that removeUndefinedValues is set to true for the option of the client? I'd need to make sure and refresh my understanding of that behavior because I dont want it to have any unwanted side effects - I know I had to deal with one in the past

Reply:
I'm worried that if you don't set it to true, then if you pass in a record with an undefined value an error will be thrown. I thought I saw potentially undefined values in the IdempotencyRecord class, so I wanted to be sure. It'd probably be best to run or add a test for it (not sure if it blows up in marshalling or in dynamo)

Reply:
Are you talking in general or for that specific scenario? In the current iteration, I believe the fields we are concerned with around the status, expiryTimestamp, idempotencyKey should always be defined by us when we first save the idempotency record into dynamo

Reply:
Both? It seems the IdempotencyRecord accepts undefined values for properties in the constructor, which implies that our code can handle it.

Reply:
Yes, our code will assume you don't want to expire if it's undefined. I think everywhere where this value will be read, it checks if its undefined before using it. In terms of how its unmarshalled, we can decide to either remove it or set it to false, but I don't think it ultimately makes a difference to how our code uses it.

Reply:
I'm not worried about the code not checking if it's undefined, my original comment was my concern about an unintended error being thrown when passing the record into a lib-dynamo method. I believe it does implicit marshalling into attribute values, which I'm worried could blow up if you have a property set to undefined.

More researching/testing would be needed as to whether you needed the flag linked in my original comment

Reply:
Good catch 

GIving the client an undefined value will result in the following result
![image](https://user-images.githubusercontent.com/111440293/194641054-d5d3db31-4209-4b80-807e-a490aedfa112.png)